File: Amazon/Onsite/C++/Next Greater Element I.cpp

/*
題意: find num1 in nums2 array's first greater number (Caution : the order of nums2 does not change)

Input: nums1 = [4,1,2], nums2 = [1,3,4,2].
Output: [-1,3,-1]
Explanation:
    For number 4 in the first array, you cannot find the next greater number for it in the second array, so output -1.
    For number 1 in the first array, the next greater number for it in the second array is 3.
    For number 2 in the first array, there is no next greater number for it in the second array, so output -1.


Solution:
	Monotonic Decreasing Stack --> We traverse num2, if traversed current number is greater than nums at stack top, then all numbers in the stack that is smaller than current number, has current number as next greater number. 
	Consider a case : [ 5, 4, 3, 2, 1, 6] for everything < 6, their next greater value is 6. 
	
	Time Complexity : O(N) , where N as size of nums2
	Space Complexity : O(N)  worse case stack depth = N
*/
class Solution {
public:
	vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2){
		stack<int> stk;
		unordered_map<int, int> next; //key: curr, value : next greater number of curr
		for(auto& num : nums2){
			while(!stk.empty() && stk.top() < num){
				next[stk.top()] = num; stk.pop();
			}
			stk.push(num);
		}

		for(int i = 0; i < nums1.size(); ++i) {
			nums1[i] = next.find(nums[i]) == next.end() ? -1 : next[nums1[i]];
		}
		
		return nums1; 
	}
};

File: Amazon/Onsite/Python/NextGreaterElementI.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # create 单调递减栈 for nums2, 
        # [1,3,4,2] 
        # for 2: the stack is empty, d[2] = -1, push 2 to stack
        # for 4: the stack top is 2, smaller than 4, pop it out.the stack is empty, d[4] = -1, push 4 to stack
        # for 3: the stack top is 4, larger than 3, d[4] = 3, push 3 to stack, 
        # for 1: the stack top is 3, larger than 1, d[1] = 3, push 1 to stack,# [1,3,4,2] 
        d = {}
        n = len(nums2)
        stack = []
        for i in range(n-1, -1, -1):
            # 把不满足单调递栈条件的元素pop 出来
            while stack and stack[-1] <= nums2[i]:
                stack.pop()
            if len(stack) == 0:
                d[nums2[i]] = -1
            else: # 栈顶就是要找的答案
                d[nums2[i]] = stack[-1] 
            stack.append(nums2[i])
            
        res = []
        for c in nums1:
            res.append(d[c])
        return res